### PR TITLE
[Feature] Refactor rds instance v3 and read replica instance

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
1. The amount of automatically generated codes are huge and difficult to read.
2. The acc test contains invalid flavor_id.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
```
NONE
```

**Special notes for your reviewer**:

**Release note**:

```release-note
1. Replace automatically built codes with new codes.
2. Reorder schema.
3. Refactor the function of waiting state.
    After the creation function is completed, the rds instance will remain active for a few seconds and then enter the backup state. 
    It will lead to rds instance backup blocking subsequent API interface calls in some cases.
    Therefore, I divide the wait state into two steps: wait for the instance to enter the backup state (In creation function) and wait for the instance to complete the backup (In updation function).
4. Combine the public methods of the two resources.
5. Use instanceID instead of d.Id() call to reduce the execution of judgment logic.
6. Update the invalid flavor in the acc test (from 'rds.pg.c2.large' to 'rds.pg.c6.large.2' ...).
```

## PR Checklist

* [x] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed
[Rds instance] Basic, update and import
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccRdsInstanceV3_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccRdsInstanceV3_basic -timeout 360m -parallel 4
=== RUN   TestAccRdsInstanceV3_basic
=== PAUSE TestAccRdsInstanceV3_basic
=== CONT  TestAccRdsInstanceV3_basic
--- PASS: TestAccRdsInstanceV3_basic (804.97s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       805.019s
```
[Rds instance] Enterprise project id
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccRdsInstanceV3_withEpsId'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccRdsInstanceV3_withEpsId -timeout 360m -parallel 4
=== RUN   TestAccRdsInstanceV3_withEpsId
=== PAUSE TestAccRdsInstanceV3_withEpsId
=== CONT  TestAccRdsInstanceV3_withEpsId
--- PASS: TestAccRdsInstanceV3_withEpsId (690.05s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       690.100s
```
[Rds instance] Basic, update and import
```
make testacc TEST=
'./huaweicloud' TESTARGS='-run=TestAccRdsReadReplicaInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccRdsReadReplicaInstance_basic -timeout 360m -parallel 4
=== RUN   TestAccRdsReadReplicaInstance_basic
=== PAUSE TestAccRdsReadReplicaInstance_basic
=== CONT  TestAccRdsReadReplicaInstance_basic
--- PASS: TestAccRdsReadReplicaInstance_basic (1301.45s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       1301.553s
```
[Rds instance] Enterprise project id
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccRdsReadReplicaInstance_withEpsId'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccRdsReadReplicaInstance_withEpsId -timeout 360m -parallel 4
=== RUN   TestAccRdsReadReplicaInstance_withEpsId
=== PAUSE TestAccRdsReadReplicaInstance_withEpsId
=== CONT  TestAccRdsReadReplicaInstance_withEpsId
--- PASS: TestAccRdsReadReplicaInstance_withEpsId (1053.73s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud  1053.803s
```

### CODE REVIEW

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 1]
```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+		FixedIp:             d.Get("fixed_ip").(string),
+		EnterpriseProjectId: GetEnterpriseProjectID(d, config),
+		Region:              region,
+	}
 
-	arrayIndex := map[string]int{
-		"backup_strategy": 0,
-		"db":              0,
-		"volume":          0,
+	// var azList []string
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: this line should be deleted

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 2]
```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+	azList := make([]string, len(d.Get("availability_zone").([]interface{})))
+	for i, az := range d.Get("availability_zone").([]interface{}) {
+		azList[i] = az.(string)
 	}
+	createRdsOpts.AvailabilityZone = strings.Join(azList, ",")
 
-	params, err := buildRdsInstanceV3CreateParameters(opts, arrayIndex)
-	if err != nil {
-		return fmt.Errorf("Error building the request body of api(create), err=%s", err)
+	if hasFilledOpt(d, "db") {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: as `db` is a required argument, hasFilledOpt will always be true.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 3]
```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+			Version: d.Get("db.0.version").(string),
+		}
+		createRdsOpts.Datastore = &db
+		createRdsOpts.Password = d.Get("db.0.password").(string)
+		createRdsOpts.Port = strconv.Itoa(d.Get("db.0.port").(int))
 	}
-	r, err := sendRdsInstanceV3CreateRequest(d, params, client)
-	if err != nil {
-		return fmt.Errorf("Error creating RdsInstanceV3, err=%s", err)
+	if hasFilledOpt(d, "volume") {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: ditto

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 4]
```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+
+	log.Printf("[DEBUG] Waiting for instance (%s) to become running", d.Id())
+	// After the rds instance is created, it will enter the backup state.
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"BUILD", "ACTIVE"},
+		Target:     []string{"BACKING UP"},
+		Refresh:    InstanceV3StateRefreshFunc(client, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: please use `PollInterval` instead.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 5]
```diff
@@ -255,106 +232,259 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+		return fmt.Errorf("Error creating sdk client, err=%s", err)
+	}
+	page, err := instances.List(client, instances.ListRdsInstanceOpts{Id: d.Id()}).AllPages()
+	if err != nil {
+		return fmt.Errorf("Error fetching rds instances %s from HuaweiCloud: %s", d.Id(), err)
+	}
+	rdsList, err := instances.ExtractRdsInstances(page.(instances.RdsPage))
+	if len(rdsList.Instances) != 1 {
+		return fmt.Errorf("The rds instance is not exist")
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: you can use `getRdsInstanceByID`

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: Comment code deleted

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: hasFilledOpt is removed.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: hasFilledOpt is removed.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: this state check is moving from creation function to updation function.
timeout set with 15 second.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: All rds List function are replaced with `getRdsInstanceByID` and use `reflect.DeepEqual` to check whether instance is empty.

File: huaweicloud/resource_huaweicloud_rds_read_replica_instance_test.go [hunk 6]
```diff
@@ -38,122 +52,68 @@ func TestAccRdsReadReplicaInstance_basic(t *testing.T) {
...
-			return fmt.Errorf("resource %s does not exist", n)
-		}
-
-		instance = found
-		return nil
-	}
-}
-
-func testAccReadRdsReplicaInstanceBasic(val string) string {
+func testAccReadRdsReplicaInstanceV3_base(val string) string {
```
Comment (Jason-Zhang9309) [inline huaweicloud/resource_huaweicloud_rds_read_replica_instance_test.go]: suggest use more meaningful variable name

Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_read_replica_instance_test.go]: `testAccReadRdsReplicaInstanceV3_base` is used as the basis of configuration of `resource_huaweicloud_rds_read_replica_instance_test.go`, and its role and naming are consistent with the rules of `testAccRdsInstanceV3_base` in `resource_huaweicloud_rds_instance_v3_test.go`.
So, I think it's fine.

Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_read_replica_instance_test.go]: `testAccReadRdsReplicaInstanceV3_base` is the configuration basis of `testAccReadRdsReplicaInstanceV3_basic`, `testAccReadRdsReplicaInstanceV3_update` and `testAccReadRdsReplicaInstanceV3_withEpsId`.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 7]
```diff
@@ -251,78 +246,66 @@ func resourceRdsInstanceV3() *schema.Resource {
...
+
+	res, err := instances.Create(client, createRdsOpts).Extract()
 	if err != nil {
-		return fmt.Errorf("Error building the request body of api(create), err=%s", err)
-	}
-	r, err := sendRdsInstanceV3CreateRequest(d, params, client)
-	if err != nil {
-		return fmt.Errorf("Error creating RdsInstanceV3, err=%s", err)
+		return err
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: please using fmt.Errorf to add more information about the error.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go [hunk 8]
```diff
@@ -334,1128 +317,281 @@ func resourceRdsInstanceV3Create(d *schema.ResourceData, meta interface{}) error
...
-			if val[k] != v {
-				bingo = false
-				break
-			}
+	az1 := instance.Nodes[0].AvailabilityZone
+	if strings.HasSuffix(d.Get("flavor").(string), ".ha") {
+		if ok, err := isEmptyValue(reflect.ValueOf(instance.Nodes[1].AvailabilityZone)); err != nil {
+			return fmt.Errorf("Error fetching availability zone form rds instance %s, err = %s", instanceID, err)
+		} else if ok {
+			return fmt.Errorf("Rds instance %s is Ha mode but number of availability zone get one", instanceID)
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_rds_instance_v3.go]: please make the error message more clear

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: `PollInterval` is used, set with 15 seconds. Is 20 seconds will be better?
The average creation time is 4 to 5 minutes.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: Unnecessary logic and error message has been cleared.

File: huaweicloud/resource_huaweicloud_rds_instance_v3.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_rds_instance_v3.go]: finished

**Pull request merged**